File: 0118-pascals-triangle/0118-pascals-triangle.cpp
// time:o(n^2)
// space: o(n^2)
class Solution {
public:
    vector<vector<int>> generate(int numRows) {
        
        vector<vector<int>> result;
        
        for(int i=0;i<numRows;i++){
            vector<int> row(i+1 , 1);
            for(int j=1;j<i;j++){
                row[j] = result[i-1][j-1] + result[i-1][j];
            }
            result.push_back(row);
            
        }
        return result;
        
    }
};

/*
int main() {
    Solution solution;

    int numRows = 5;
    vector<vector<int>> result = solution.generate(numRows);

    // Print the generated Pascal's Triangle
    for (const auto& row : result) {
        for (int num : row) {
            cout << num << " ";
        }
        cout << endl;
    }

    return 0;
}
*/

File: 0118-pascals-triangle/0118-pascals-triangle.py
class Solution:
    def generate(self, numRows: int) -> List[List[int]]:
        
        ##create an answer 
        result = [[1]]
        
        ##iterate through number of rows i.e numsRows
        for i in  range(numRows - 1):
            ##add a zero in the beginning and end of the row and store it into a temp 
            temp = [0] + result[-1] + [0]
            ## create an empty row
            row = []
            ## building the next row = length of previous row + 1
            for j in range(len(result[-1]) + 1 ):
                ## temp[j] and tep[j+1] are two pointers 
                row.append(temp[j] + temp[j+1]) 
            result.append(row)
            
        return result   
        